gr-digital/include/gnuradio/digital/lfsr.h

```diff
@@ -78,32 +81,33 @@ namespace digital {
 class lfsr
 {
 private:
-    uint32_t d_shift_register;
-    uint32_t d_mask;
-    uint32_t d_seed;
-    uint32_t d_shift_register_length; // less than 32
-
-    static uint32_t popCount(uint32_t x)
-    {
-        uint32_t r = x - ((x >> 1) & 033333333333) - ((x >> 2) & 011111111111);
-        return ((r + (r >> 3)) & 030707070707) % 63;
-    }
+    uint64_t d_shift_register;
+    uint64_t d_mask;
+    uint64_t d_seed;
+    uint8_t d_shift_register_length; // less than 64
 
 public:
-    lfsr(uint32_t mask, uint32_t seed, uint32_t reg_len)
+    lfsr(uint64_t mask, uint64_t seed, uint8_t reg_len)
         : d_shift_register(seed),
           d_mask(mask),
           d_seed(seed),
           d_shift_register_length(reg_len)
     {
-        if (reg_len > 31)
-            throw std::invalid_argument("reg_len must be <= 31");
+        if (reg_len > 63)
+            throw std::invalid_argument("reg_len must be <= 63");
     }
 
     unsigned char next_bit()
     {
         unsigned char output = d_shift_register & 1;
-        unsigned char newbit = popCount(d_shift_register & d_mask) % 2;
+        uint64_t newbit;
+#ifdef HAVE_BUILTIN_PARITYL
+        newbit = __builtin_parityl(d_shift_register & d_mask);
```

Comment:
From [1]
> I don't have GR set up on a RPi, but here's the code generated by a test of __builtin_parity() and __builtin_parityl() on a RPi3. They both call __paritysi2() which looks no less efficient than calling popCount().

`__builtin_parity()` (or `__builtin_popcount()`) will not issue special instructions unless they are enabled: `-mpopcnt` for x86/x64 (you might get away with one of the `-msse4X` options, but that depends on AMD vs. Intel as far as I remember).

From [2]
> Built in functions surely aren't portable, right?

The resulting binary will not be portable if special instructions are issued (no runtime detection), but the source is (at least when using GCC/Clang).

[1] https://github.com/gnuradio/gnuradio/pull/3968#issuecomment-734927631
[2] https://github.com/gnuradio/gnuradio/pull/3968#issuecomment-735467759

Reply:
[1]: I'm looking for the use of the PF flag, and maybe it's the setnp instruction in x64? I seem to get this with __builtin_parityl without having to enable anything when compiling GR. I get a speedup of 50% on my system. Are you sure PF and setnp depend on SSE or others? From what I understand this asm-instruction was included in the original x86 set.

[2]: Do you know of a better way to solve this? When I tried the VOLK-way, the overhead gave me the same performance as previous.

Reply:
[1]: To be honest, this is the first time I came across `setnp` (seems to be indeed part of the original i386 instruction set) and the Parity Flag (PF). `setnp` alone ist not sufficient as it processes a single byte only. This is the code produced by GCC with `__builtin_parity(val)` if `popcnt` is not available: (`RCX` contains `val`, result is in `RAX`)
```
0x55555555518e  <+   30>        48 89 ce              mov    %rcx,%rsi
0x555555555191  <+   33>        89 cf                 mov    %ecx,%edi
0x555555555193  <+   35>        48 c1 ee 20           shr    $0x20,%rsi
0x555555555197  <+   39>        31 f7                 xor    %esi,%edi
0x555555555199  <+   41>        89 f8                 mov    %edi,%eax
0x55555555519b  <+   43>        c1 ef 10              shr    $0x10,%edi
0x55555555519e  <+   46>        31 f8                 xor    %edi,%eax
0x5555555551a0  <+   48>        30 e0                 xor    %ah,%al
0x5555555551a2  <+   50>        0f 9b c0              setnp  %al
```
[2] Short of advising users to compile with `-march=native`, no. Using `volk` here is pointless because only a single value is processed.